File: Lecture09/075_adding_operations_or_variants.java
// Programming Languages
// Lecture 09: Adding Operations or Variants

abstract class Exp {
    public abstract Value eval();
    public abstract String toString();
    public abstract boolean hasZero();
    public abstract Exp noNegConstants();
}

abstract class Value extends Exp {

}

class Int extends Value {
    protected int i;

    public Int(int i) {
        this.i = i;
    }

    @Override
    public Value eval() {
        return this;
    }

    @Override
    public String toString() {
        return "" + i;
    }

    @Override
    public boolean hasZero() {
        return i == 0;
    }

    @Override
    public Exp noNegConstants() {
        if (i < 0) {
            return new Negate(new Int(-i));
        } else {
            return this;
        }
    }
}

class Negate extends Exp {
    private Exp e;

    public Negate(Exp e) {
        this.e = e;
    }

    @Override
    public Value eval() {
        return new Int(-((Int)(e.eval())).i);
    }

    @Override
    public String toString() {
        return "-(" + e.toString() + ")";
    }

    @Override
    public boolean hasZero() {
        return e.hasZero();
    }

    @Override
    public Exp noNegConstants() {
        return new Negate(e.noNegConstants());
    }
}

class Add extends Exp {
    private Exp e1;
    private Exp e2;
    
    public Add(Exp e1, Exp e2) {
	    this.e1 = e1;
	    this.e2 = e2;
    }

    @Override
    public Value eval() {
	    return new Int(((Int)(e1.eval())).i + ((Int)(e2.eval())).i);
    }
    
    @Override
    public String toString() {
	    return "(" + e1.toString() + " + " + e2.toString() + ")";
    }
    
    @Override
    public boolean hasZero() {
	    return e1.hasZero() || e2.hasZero();
    }

    @Override
    public Exp noNegConstants() {
        return new Add(e1.noNegConstants(), e2.noNegConstants());
    }
}

class Mult extends Exp {
    private Exp e1;
    private Exp e2;
    
    public Mult(Exp e1, Exp e2) {
        this.e1 = e1;
        this.e2 = e2;
    }

    @Override
    public Value eval() {
        return new Int(((Int)(e1.eval())).i * ((Int)(e2.eval())).i);
    }
    
    @Override
    public String toString() {
        return "(" + e1.toString() + " * " + e2.toString() + ")";
    }
    
    @Override
    public boolean hasZero() {
        return e1.hasZero() || e2.hasZero();
    }

    @Override
    public Exp noNegConstants() {
        return new Mult(e1.noNegConstants(), e2.noNegConstants());
    }
}

final class Main {
    public static void main(String[] args) {
        // 5 + 10 * (-3)
        Exp e = new Add(new Int(5), new Mult(new Int(10), new Negate(new Int(3))));
        System.out.println("Does e have a zero? " + e.hasZero());
        System.out.println(e.toString() + " = " + e.eval().toString());

        // 5 + 10 * (-3)
        Exp e2 = new Add(new Int(5), new Mult(new Int(10), new Int(-3))).noNegConstants();
        System.out.println("Does e2 have a zero? " + e2.hasZero());
        System.out.println(e2.toString() + " = " + e2.eval().toString());
    }
}

File: Lecture09/075_adding_operations_or_variants.cpp
// Programming Languages
// Lecture 09: Adding Operations or Variants

#include <iostream>
#include <memory>
#include <string>

class Value;

class Exp
{
public:
    virtual std::shared_ptr<Value> eval() noexcept = 0;

    virtual std::string to_string() const noexcept = 0;

    virtual bool has_zero() const noexcept = 0;

    virtual std::shared_ptr<Exp> no_neg_constants() noexcept = 0;
};

class Value: public Exp
{

};

class Int: public Value
{
public:
    Int(int _i) noexcept
        : i{_i}
    {
        // Empty
    }

    std::shared_ptr<Value> eval() noexcept override
    {
        return std::make_shared<Int>(i);
    }

    std::string to_string() const noexcept override
    {
        return std::to_string(i);
    }

    bool has_zero() const noexcept override
    {
        return i == 0;
    }

    std::shared_ptr<Exp> no_neg_constants() noexcept override;

public:
    int i;
};

class Negate: public Exp
{
public:
    Negate(std::shared_ptr<Exp> _e) noexcept
        : e{_e}
    {
        // Empty
    }

    std::shared_ptr<Value> eval() noexcept override
    {
        return std::make_shared<Int>(-std::dynamic_pointer_cast<Int>(e->eval())->i);
    }

    std::string to_string() const noexcept override
    {
        return "-(" + e->to_string() + ")";
    }

    bool has_zero() const noexcept override
    {
        return e->has_zero();
    }

    std::shared_ptr<Exp> no_neg_constants() noexcept override
    {
        return std::make_shared<Negate>(e->no_neg_constants());
    }

private:
    std::shared_ptr<Exp> e;
};

std::shared_ptr<Exp> Int::no_neg_constants() noexcept
{
    if (i < 0)
    {
        return std::make_shared<Negate>(std::make_shared<Int>(-i));
    }
    else
    {
        return std::make_shared<Int>(i);
    }
}

class Add: public Exp
{
public:
    Add(std::shared_ptr<Exp> _e1, std::shared_ptr<Exp> _e2) noexcept
        : e1{_e1}, e2{_e2}
    {
        // Empty
    }

    std::shared_ptr<Value> eval() noexcept override
    {
        return std::make_shared<Int>(std::dynamic_pointer_cast<Int>(e1->eval())->i + std::dynamic_pointer_cast<Int>(e2->eval())->i);
    }

    std::string to_string() const noexcept override
    {
        return "(" + e1->to_string() + " + " + e2->to_string() + ")";
    }

    bool has_zero() const noexcept override
    {
        return e1->has_zero() || e2->has_zero();
    }

    std::shared_ptr<Exp> no_neg_constants() noexcept override
    {
        return std::make_shared<Add>(e1->no_neg_constants(), e2->no_neg_constants());
    }

private:
    std::shared_ptr<Exp> e1;
    std::shared_ptr<Exp> e2;
};

class Mult: public Exp
{
public:
    Mult(std::shared_ptr<Exp> _e1, std::shared_ptr<Exp> _e2) noexcept
        : e1{_e1}, e2{_e2}
    {
        // Empty
    }

    std::shared_ptr<Value> eval() noexcept override
    {
        return std::make_shared<Int>(std::dynamic_pointer_cast<Int>(e1->eval())->i * std::dynamic_pointer_cast<Int>(e2->eval())->i);
    }

    std::string to_string() const noexcept override
    {
        return "(" + e1->to_string() + " * " + e2->to_string() + ")";
    }

    bool has_zero() const noexcept override
    {
        return e1->has_zero() || e2->has_zero();
    }

    std::shared_ptr<Exp> no_neg_constants() noexcept override
    {
        return std::make_shared<Mult>(e1->no_neg_constants(), e2->no_neg_constants());
    }

private:
    std::shared_ptr<Exp> e1;
    std::shared_ptr<Exp> e2;
};

int main()
{
    // 5 + 10 * -3)
    auto e = std::make_shared<Add>(
        std::make_shared<Int>(5),
        std::make_shared<Mult>(
            std::make_shared<Int>(10),
            std::make_shared<Negate>(std::make_shared<Int>(3))
        )
    );

    std::cout << std::boolalpha;
    std::cout << "Does e have zero? " << e->has_zero() << "\n";
    std::cout << e->to_string() << " = " << e->eval()->to_string() << "\n";

    // 5 + 10 * (-3)
    auto e2 = std::make_shared<Add>(
        std::make_shared<Int>(5),
        std::make_shared<Mult>(
            std::make_shared<Int>(10),
            std::make_shared<Int>(-3)
        )
    )->no_neg_constants();

    std::cout << "Does e2 have zero? " << e2->has_zero() << "\n";
    std::cout << e2->to_string() << " = " << e2->eval()->to_string() << "\n";

    return EXIT_SUCCESS;
}